src/menu/Menu.ts

```diff
@@ -0,0 +1,311 @@
+
+		if (!disabled && expandOnClick) {
+			this._toggleDisplay();
+		}
+	}
+
+	onLabelKeypress(event: KeyboardEvent) {
+		const { disabled } = this.properties;
+
+		if (!disabled && event.key === 'Enter') {
+			this._toggleDisplay();
+		}
+	}
+
+	onMenuFocus() {
+		const { disabled, hidden = this._getDefaultHidden() } = this.properties;
+		if (!disabled && hidden) {
+			const onRequestShow = this.properties.onRequestShow;
+			onRequestShow && onRequestShow();
+		}
+	}
+
+	onMenuMouseEnter() {
+		const {
+			disabled,
+			expandOnClick = true
+		} = this.properties;
+
+		if (!disabled && !expandOnClick) {
+			this._hideTimer && this._hideTimer.destroy();
+			this._toggleDisplay(true);
+		}
+	}
+
+	onMenuMouseLeave() {
+		const {
+			disabled,
+			expandOnClick = true,
+			hideDelay = 300
+		} = this.properties;
+
+		if (!disabled && !expandOnClick) {
+			this._hideTimer && this._hideTimer.destroy();
+			if (hideDelay) {
+				this._hideTimer = createTimer(() => {
+					this._toggleDisplay(false);
+				}, hideDelay);
+				this.own(this._hideTimer);
+			}
+			else {
+				this._toggleDisplay(false);
+			}
+		}
+	}
+
+	render(): DNode {
+		const {
+			hidden = this._getDefaultHidden(),
+			id = uuid(),
+			nested,
+			role = 'menu'
+		} = this.properties;
+
+		const label = this.renderLabel(id);
+		const menu = v('nav', {
+			id,
+			role,
+			classes: this.classes.apply(this, this._getMenuClasses(hidden, Boolean(label))),
+			afterCreate: this._afterCreate,
+			afterUpdate: this._afterUpdate,
+			onfocusin: this.onMenuFocus
+		}, this.children);
+
+		if (label) {
+			return v('div', {
+				classes: this.classes(css.container, nested ? css.nestedMenuContainer : null),
+				onmouseenter: this.onMenuMouseEnter,
+				onmouseleave: this.onMenuMouseLeave
+			}, [ label, menu ]);
+		}
+
+		return menu;
+	}
+
+	renderLabel(id: string): DNode | void {
+		const { disabled, label, overrideClasses } = this.properties;
+
+		if (label) {
+			const properties = typeof label === 'string' ? { label } : label;
+			return w(MenuItem, assign({
+				disabled,
+				getAriaProperties: this._getLabelAriaProperties.bind(this, id),
+				hasMenu: true,
+				overrideClasses: overrideClasses || css,
+				onClick: this.onLabelClick,
+				onKeypress: this.onLabelKeypress
+			}, properties));
+		}
+	}
+
+	protected _afterCreate(element: HTMLElement) {
+		const { animate = true } = this.properties;
+		this._initialRender = false;
+
+		if (animate) {
+			const hidden = element.classList.contains(css.hidden);
+			this._wasOpen = !hidden;
+
+			if (hidden) {
+				element.style.height = '0';
+			}
+		}
+	}
+
+	protected _afterUpdate(element: HTMLElement) {
+		const { animate = true, hidden = this._getDefaultHidden(), label } = this.properties;
+
+		if (!label) {
+			return;
+		}
+
+		if (!animate) {
+			// In case `animate` was previously `true`, remove any `height` property set on the node.
+			element.style.height = null;
+			return;
+		}
+
+		if (typeof animate === 'function') {
```

Comment:
Are we sure this will work correctly since opening and closing is only occurring after updates, not after creation? I think it's possible that some combination of upstream property changes could trigger the Menu to be recreated vs. updated, and if `hidden` changes as part of those upstream property changes, opening and closing should still occur.